### PR TITLE
SURFconext auth module and nginx integration

### LINKED ISSUE #50: Login with SurfConext
## Requirements: 

* When clicking on the sign-in button, open a modal with authentication access (SSO - Single sign options). 
* Once is logged in, remove the modal. (the user stays where it was). 
* Landing page vs user profile page: when the anonim user is in the landing page ("/"), after authenticating, redirect the landing page to their user profile ("/<username>")
* If the user is authenticated: do not show the [sign in] button, but the avatar with the hover menu with: [sign out]
image

## Nice to have
- When authenticating with the provider, do not redirect the application to their page, but show their credentials page into a floating window. (**popup mode**)

## Wireframe
This is a wireframe, don't follow strictly the styles on it, they will be defined later on with the definite UI design

### PR DESCRIPTION
**Pull request by dmijatovic**
# SURFconext auth module and nginx integration

Fixes #50 #8 

The implementation of SURFConext authentication on the front-end and the integration with nginx as reverse proxy.

How to test:

* `docker-compose down --volumes`: to remove old containers
* `docker-compose build`: to build new images
* `docker-compose up`: to start the solution. Examine the logs: the test should be passing for fronend and auth-tests
* The app should be visible on http://localhost
* Test data migration module by following these steps: 
```bash
# navigate to data-migration folder
cd data-migration
# run docker-compose
docker-compose up
```
* Test login with SURFconext:
    * First navigate to [software page](http://localhost/software). Note! you should be redirected back to this page after login process is completed
    * click on SingIn button in the top right corner of the screen 
image
    * select SURFconext Test IdP and login with credentials: professor1 / professor1
    * you should be redirect back to software page and the avatar should be visible in the right top corner 
image
    
*  Test refresh tokens with SURFconext: navigate to my profile page using menu at user avatar. Note the exp value and wait for about 1 minute. The exp value will change. You can also examine browser console logs which should have logs about refresh schedule. Note! The token expiration time is currently set to very low value in order to test token refresh is working properly. The token expiration time should be 

## Nginx setup

The reverse proxy is setup with this routes:
* [root](http://localhost): home page of RSD (frontend)
* [auth](http://localhost/auth): authentication module (Java)
* [api/v1](http://localhost/api/v1): postgREST backend service (Haskel)
* [api/fe](http://localhost/api/fe): api service of frontend (currently citations download and refresh tokens endpoint) (NodeJS)

### CODE REVIEW

File: authentication/src/main/java/nl/esciencecenter/rsd/authentication/JwtCreator.java [hunk 1]
```diff
@@ -10,7 +10,8 @@
 
 public class JwtCreator {
 
-	static final long ONE_HOUR_IN_MILLISECONDS = 3600_000L; // 60 * 60 * 1000
+	// static final long ONE_HOUR_IN_MILLISECONDS = 3600_000L; // 60 * 60 * 1000
+	static final long ONE_HOUR_IN_MILLISECONDS = 360_000L; // 6 * 60 * 1000
```
Comment (ewan-escience) [inline authentication/src/main/java/nl/esciencecenter/rsd/authentication/JwtCreator.java:14]: What happened here?

File: authentication/src/main/java/nl/esciencecenter/rsd/authentication/Main.java [hunk 2]
```diff
@@ -75,7 +95,8 @@ public static void main(String[] args) throws IOException {
 	}
 
 	static void setJwtCookie(Context ctx, String token) {
-		ctx.header("Set-Cookie", "rsd_token=" + token + "; Secure; HttpOnly; Path=/; SameSite=Lax");
+		// Max-Age of 1 hour in seconds
```
Comment (ewan-escience) [inline authentication/src/main/java/nl/esciencecenter/rsd/authentication/Main.java:98]: This comment isn't implemented

File: README.md [hunk 3]
```diff
@@ -7,40 +8,39 @@ This repo contains the new RSD-as-a-service implementation
 
 ## Building
 
-The program can easily be built with `docker-compose`:
+The program can easily be built with `docker-compose`. Each service builds the image using specific version (see docker-compose.yml file). Ensure that version number is increased in docker-compose.yml file when the source code of that service is changed.
 
-1. Run the command `docker-compose build`. You might see an error from the service `data-migration_1`. This is intentional and is used as a workaround to only download dependencies if needed but to not start the program. The whole process itself should end successfully.
```
Comment (ewan-escience) [inline README.md:12]: This message should be transferred to the dedicated README.

File: docker-compose.yml [hunk 4]
```diff
@@ -21,39 +25,38 @@ services:
   backend:
     container_name: backend
     build: ./backend-postgrest
-    ports:
-      - "3500:3500"
```
Comment (ewan-escience) [inline docker-compose.yml:25]: Please undo this, for local development it should be possible to reach the backend.

File: docker-compose.yml [hunk 5]
```diff
@@ -21,39 +25,38 @@ services:
...
 
-  authentication:
-    container_name: authentication
+  auth:
+    container_name: auth
     build:
       context: .
       dockerfile: ./authentication/dockerfile
-    ports:
-      - "7000:7000"
```
Comment (ewan-escience) [inline docker-compose.yml:54]: Please undo this, for local development it should be possible to reach the backend.

File: frontend/README.md [hunk 6]
```diff
@@ -14,10 +14,13 @@ Based on the features in the legacy application and the current requirements we
 
 ### Environment variables
 
-For oAuth implementation we need number of env variables. Copy env.local.example file to `env.local` and provide values required for next-auth module.
+For oAuth implementation we need env variables. Copy env.local.example file to `env.local` and provide values required for next-auth module. In addition we use few public env variables that exposed to the browser. These values are stored in .env file. See [next documentation page for more info](https://nextjs.org/docs/basic-features/environment-variables).
```
Comment (ewan-escience) [inline frontend/README.md:17]: It looks like the configuration is now scattered over multiple files. Please undo this, configuration (including secrets) should be in one file.

File: authentication/src/main/java/nl/esciencecenter/rsd/authentication/JwtCreator.java (hunk 1 above)
Comment (dmijatovic) [reply authentication/src/main/java/nl/esciencecenter/rsd/authentication/JwtCreator.java:14]: We should reverse back to old. Good point!

File: authentication/src/main/java/nl/esciencecenter/rsd/authentication/Main.java (hunk 2 above)
Comment (dmijatovic) [reply authentication/src/main/java/nl/esciencecenter/rsd/authentication/Main.java:98]: Well spotted. I will implement max-age.

File: README.md (hunk 3 above)
Comment (dmijatovic) [reply README.md:12]: Good point. I will include it in data-migation read me.

File: frontend/README.md (hunk 6 above)
Comment (dmijatovic) [reply frontend/README.md:17]: OK. I will group all env variables into one file and update example env.local.example

File: docker-compose.yml (hunk 5 above)
Comment (dmijatovic) [reply docker-compose.yml:54]: It is using external url. I defined env variables for this purpose.

**Review by ewan-escience — APPROVED** [review]

**Pull request merged**